consensus/hotstuff/timeoutaggregator/timeout_aggregator.go

```diff
@@ -0,0 +1,231 @@
+		queuedTimeouts:             queuedTimeouts,
+	}
+
+	componentBuilder := component.NewComponentManagerBuilder()
+	for i := 0; i < defaultTimeoutAggregatorWorkers; i++ { // manager for worker routines that process inbound events
+		componentBuilder.AddWorker(func(ctx irrecoverable.SignalerContext, ready component.ReadyFunc) {
+			ready()
+			aggregator.queuedTimeoutsProcessingLoop(ctx)
+		})
+	}
+	componentBuilder.AddWorker(func(ctx irrecoverable.SignalerContext, ready component.ReadyFunc) {
+		ready()
+		aggregator.enteringViewProcessingLoop(ctx)
+	})
+
+	aggregator.ComponentManager = componentBuilder.Build()
+	return aggregator, nil
+}
+
+// queuedTimeoutsProcessingLoop is the event loop which waits for notification about pending work
+// and as soon as there is some it triggers processing.
+// All errors are propagated to SignalerContext
+func (t *TimeoutAggregator) queuedTimeoutsProcessingLoop(ctx irrecoverable.SignalerContext) {
+	notifier := t.queuedTimeoutsNotifier.Channel()
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-notifier:
+			err := t.processQueuedTimeoutEvents(ctx)
+			if err != nil {
+				ctx.Throw(fmt.Errorf("internal error processing queued timeout events: %w", err))
+				return
+			}
+		}
+	}
+}
+
+// processQueuedTimeoutEvents is a function which dispatches previously queued timeouts on worker thread
+// This function is called whenever we have queued timeouts ready to be dispatched.
+// No errors are expected during normal operations.
+func (t *TimeoutAggregator) processQueuedTimeoutEvents(ctx context.Context) error {
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		default:
+		}
+
+		msg, ok := t.queuedTimeouts.Pop()
+		if ok {
+			timeoutObject := msg.(*model.TimeoutObject)
+			err := t.processQueuedTimeout(timeoutObject)
+			if err != nil {
+				return fmt.Errorf("could not process pending TO %v: %w", timeoutObject.ID(), err)
+			}
+
+			t.log.Info().
+				Uint64("view", timeoutObject.View).
+				Str("timeout_id", timeoutObject.ID().String()).
+				Msg("TO has been processed successfully")
+
+			continue
+		}
+
+		// when there is no more messages in the queue, back to the loop to wait
+		// for the next incoming message to arrive.
+		return nil
+	}
+}
+
+// processQueuedTimeout performs actual processing of queued timeouts, this method is called from multiple
+// concurrent goroutines.
+func (t *TimeoutAggregator) processQueuedTimeout(timeoutObject *model.TimeoutObject) error {
+	// TODO: replace this check by a specific function to query if there is epoch by view
+	_, err := t.committee.LeaderForView(timeoutObject.View)
+	if err != nil {
+		// ignore TO if we don't have information for epoch
+		if errors.Is(err, model.ErrViewForUnknownEpoch) {
+			return nil
+		}
+		return fmt.Errorf("unknown error when querying epoch state for view %d: %w", timeoutObject.View, err)
+	}
+
+	collector, created, err := t.collectors.GetOrCreateCollector(timeoutObject.View)
```

Comment:
I'm wondering if we should limit how far ahead of the current view we will allow TOs to be, and still cache them. Thinking "out loud" a bit here 👇 

Epochs last about ~500,000 views currently, and we create a collector object for a view before validating the first TO for that view. So an adversary could in the worst case cause us to create ~500,000 unnecessary TimeoutCollector objects in memory which we would prune very slowly. Some thoughts:
* If this happens, it would impact memory usage but not pruning, because of the [clever pruning optimization](https://github.com/onflow/flow-go/pull/2740/files#diff-edb27b0b5326b242a3b35e845cf264382ded9b3ae2f92fd6a50dc55531c4d051R110-R113)
* To accomplish this, an adversary would need to send invalid TCs, which is a slashable offense. In the mature protocol, the adversary should be ejected before they can send enough messages
* The attack is expensive (sending messages linear with the resulting adverse memory usage increase)

Overall, I don't think we need any code changes to deal with this spamming potential, but we could add a comment about why its safe:
```suggestion
	// We create a timeout collector before validating the first TO, so processing an invalid TO will
	// result in a collector being added, until the corresponding view is pruned.
	// Since epochs on Mainnet have 500,000-1,000,000 views, there is an opportunity for memory exhaustion here.
	// However, because only invalid TOs can result in creating collectors at an arbitrary view,
	// and invalid TOs are slashable, resulting eventually in ejection of the sender, this attack is impactical.
	collector, created, err := t.collectors.GetOrCreateCollector(timeoutObject.View)
```

Reply:
good observation. Agree on no code change needed but adding documentation.